### PR TITLE
feat: Import/Export Content Item commands

### PR DESCRIPTION
**Pull request by rs-amp**
This PR adds commands for import/export. These are pretty complicated, so the testing is quite extensive but doesn't reach 100% yet on the import task.

# Export Task
`dc-cli content-item export <dir>`
The export task exports content from the configured hub into the destination folder. This content is in the format returned by the api - a wrapper object containing the metadata (eg. label, lastPublishedVersion), then the actual content under the `body` property.

By default, this command will export content from all repositories, and all folders within them, as long as they aren't archived. By using parameters, you can limit this to exporting from a specific folder, repository, and to filter items within these by name/schema id. The filters can be exact, or by regex, which can be achieved by surrounding the filter in forward slashes (like /filter.*regex/).

This command scans the content dependancies of items being exported. If they contain any items that are not included in the export scope, or are archived, then they will be exported under the folder `_dependancies/`. Dependancies are traversed recursively, so any dependancies introduced this way can also include dependancies of their own.

Parameters:
- **--repoId:** Export content from within a given repository. Directory structure will start at the specified repository. Will automatically export all contained folders.
- **--folderId:** Export content from within a given folder. Directory structure will start at the specified folder. Can be used in addition to repoId.
- **--schemaId:** Export content with a given or matching Schema ID. A regex can be provided, surrounded with forward slashes. Can be used in combination with other filters.
- **--name:** Export content with a given or matching Name. A regex can be provided, surrounded with forward slashes. Can be used in combination with other filters.
- **--logFile:** Path to a log file to write to.

# Import Task
`dc-cli content-item import <dir>`
The import task imports content from the specified directory into the configured hub. By default, this will treat folders in your directory as Repositories, and try to map them to ones of the same name in your target hub. From there, it will import content items to each repo, recreating the folder structure in each. This format is an exact match for the export format, so it is ideal for copying content within or between hubs.

Similar to the export task, you can import into a specific repo or folder. When doing this, the top level directory should just contain the content and subfolders, rather than containing individual repositories as folders. All folder structure will be created before importing content items.

If content items are imported that are missing types, or have them misconfigured, the command will be able to resolve these issues itself on user request. For example, if a content type exists in a hub, but is not assigned to a repo relevant content is being imported to, then the command can automatically create the missing assignment on user request.

At the core of the import process lies the ID mapping and the dependancy tree. When content is imported to a a new destination, it will be created with a new ID courtesy of the DC api. This will cause problems with other content that references it, as that content will now need to reference the _new_ ID for that content, which will only become available once it is imported. During import an ID mapping is built for this reason, between the IDs in the files and the new ones, and it is saved afterwards for a future use. Because of this, we need to import content in an order that ensures we already have the ID for all contained content items when importing it - a dependancy tree must be built.

The dependancy tree will follow all standard content links and references in content items, and will track dependancies and dependants for each item. It will build a list of "levels" - each level requires the content in the previous level to be imported. This way, we can import each item in order, once, and all required IDs will always be available. 

Circular dependancies do not appear on the tree: instead they are detected and placed on a separate list simply called `circularLinks`. These use a different strategy when importing - create the content first to get all the new IDs (with null or invalid references), then update it afterwards with the new IDs in the references. Circular content is not a typical use case for DC since it is unpublishable, but the case is covered here in case it is important in future.

The ID mapping is saved to a default location in the `.amplience` folder, based on the destination (hub/repo/folder, and its id). This location is printed, but you can also specify it with `--logFile`. This ID mapping is also _loaded_ on future executions of the import command, which allows you to reimport the same content by overwriting existing items, and import _new_ content with references to those you have already imported.

You have the option to import content with the `--publish` option. This will publish all content with `lastPublishedVersion` present in the metadata - so all content that was published when it was exported. This requests publishes from a "publish queue", which responsibly limits the number of concurrent publishes.

There are two key optimisations to publishing to avoid it making too many requests:
- Dependancy elimination: If content item 'A' is depended on by another content item 'B' that will be published anyways, then the publish request does not need to be made for 'A' as it will be covered when publishing 'B'. In our demo site, this roughly results in 1/3 of publishes that need to be performed.
- Don't publish when unmodified: If a content item isn't modified when it is imported with an existing mapping, then it does not need to be published again.

You can override the second item if you pass in `--republish`, for instance if you depend on some publish trigger and need it to fire.

There are 6 different situations where the import will stop and ask you if you want to continue:

- Content already exists in the mapping, asks if you want to overwrite the items rather than skipping them.
- Content Types are missing but schemas are present, asks if you want to automatically create them (or abort)
- Content Types are not assigned to target repos, asks if you want to automatically assign them (or abort)
- Content Type Schema are missing, asks if you want to skip content items that use it. (and informs you which, how many)
- Referenced Content Items (dependancies) are missing, asks if you are ok with setting them null or skipping them. (or abort)
- Repositories in the import directory could not be mapped to ones in DC. Asks if you are ok with skipping them. (or abort)

Parameters:
- **--baseRepo:** Import matching the given repository to the import base directory, by ID. Folder structure will be followed and replicated from there.
- **--baseFolder:** Import matching the given folder to the import base directory, by ID. Folder structure will be followed and replicated from there.
- **--mapFile:** Mapping file to use when updating content that already exists. Updated with any new mappings that are generated. If not present, will be created.
- **-f, --force:** Overwrite content, create and assign content types, and ignore content with missing types/references without asking.
- **-v, --validate:** Only recreate folder structure - content is validated but not imported.
- **--skipIncomplete:** Skip any content items that has one or more missing dependancy.
- **--publish:** Publish any content items that have an existing publish status in their JSON.
- **--republish:** Republish content items regardless of whether the import changed them or not. (--publish not required)
- **--logFile:** Path to a log file to write to.

# Publish Queue
The import task requires the ability to automatically publish content items. The responsibility of keeping a maximum number of concurrent publishes is on us, since there is no support in the management sdk.

The interface for the queue is very simple. All the user has to do is request publishes for specific content items, then wait for them all to complete afterwards. Each publish calls a helper function called "rateLimit", which regulates how many concurrent publishes can be active at once.

The rate limit is currently configured to 10 concurrent items being published. When 10 publishes have started, the rate limit checks to see if the oldest publish has complete. If not, then it waits for it to complete before continuing, by polling the processing job status every second. This is enforced so that we can't start any new publishes until older ones complete, so at any given time we can only have a set number of concurrent publishes in progress. Because of this, publish start can take a variable amount of time, and to wait for _all_ started publishes to complete, you will need to `waitForAll` at the end.

# Import-Revert
This is not used yet, but it will be used in future for the copy/move tasks and is fully tested.

### CODE REVIEW

File: src/common/import/publish-queue.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+    this.waitInProgress = true;
+
+    const oldestJob = this.inProgressJobs[0];
+    this.inProgressJobs.splice(0, 1);
+
+    // Request the status for the oldest ID.
+    // If it's still not published/errored, then wait a bit and try again.
+
+    while (true) {
```
Comment (Benj0s) [inline src/common/import/publish-queue.ts:98]: As discussed previously; we should limit the number of retries here

**Review by Benj0s — APPROVED** [review]

File: src/commands/content-type-schema/__mocks__/readline.ts [hunk 2]
```diff
@@ -3,6 +3,7 @@ let responseQueue: string[] = [];
 module.exports = {
   createInterface: jest.fn().mockReturnValue({
     question: jest.fn().mockImplementation((questionText, cb) => {
+      console.log(questionText);
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply src/commands/content-type-schema/__mocks__/readline.ts:6]: It's to log the question in the console, with the other logs the program normally does. Jest suppresses these, but you can see them when running/debugging the test individually so it's still useful to have.

File: src/common/content-item/content-dependancy-tree.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    this.levels = stages;
+    this.circularLinks = info;
+    this.all = allInfo;
+    this.byId = new Map(allInfo.map(info => [info.owner.content.id as string, info]));
+    this.requiredSchema = Array.from(requiredSchema);
+  }
+
+  public searchObjectForContentDependancies(
+    item: RepositoryContentItem,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
[the remark this answers is not in the document]
Comment (rs-amp) [reply src/common/content-item/content-dependancy-tree.ts:101]: The input to this function is a content body, but on recursive calls it is an object or array.

**Comment by rs-amp** [comment]
Addressed feedback from Ben.

**Comment by rs-amp** [comment]
I've made a few changes for hopefully the final pass:

- add --publish flag for export
- use cached schemas for top level type validation
- publishes now perform sequentially rather than concurrently

File: src/commands/content-item/export.ts [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+    baseFolder = false;
+    processFolders = nextFolders.splice(0, Math.min(nextFolders.length, parallelism));
+  }
+
+  if (publish) {
+    for (let i = 0; i < items.length; i++) {
+      const item = items[i];
+
+      const publishedVersion: number | undefined = (item.item as any).lastPublishedVersion;
```
Comment (easen-amp) [inline src/commands/content-item/export.ts:200]: You can use `ContentItem` instead of `any`

File: src/common/content-item/amplience-schema-validator.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+  constructor(private schemaLookup: (uri: string) => Promise<ContentTypeSchema | undefined>) {
+    const ajv = new Ajv({
+      loadSchema: this.loadSchema.bind(this),
+      unknownFormats: ['symbol', 'color', 'markdown', 'text'],
+      schemaId: 'auto'
+    });
+
+    // eslint-disable-next-line @typescript-eslint/no-var-requires
+    const draft4 = require('ajv/lib/refs/json-schema-draft-04.json');
```
Comment (easen-amp) [inline src/common/content-item/amplience-schema-validator.ts:45]: DC currently supports draft 4 and draft 7 as well as some bespoke settings that are not in the json-schema spec. This is normally maintained/managed server side, allowing us to add new features/fixes.

Do you require client side validation? Or is it an optimisation feature to speed up the import phase?

File: src/common/content-item/content-dependancy-tree.ts [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+export const referenceTypes = [
+  'http://bigcontent.io/cms/schema/v1/core#/definitions/content-link',
+  'http://bigcontent.io/cms/schema/v1/core#/definitions/content-reference'
+];
+
+type RecursiveSearchStep = Body | ContentDependancy | Array<Body>;
+
+export class ContentDependancyTree {
```
Comment (easen-amp) [inline src/common/content-item/content-dependancy-tree.ts:43]: I can't see any unit tests for this class

Comment (easen-amp) [reply src/common/content-item/content-dependancy-tree.ts:43]: After running the tests I can see there is coverage for this class, so I think it's indirectly tested from some of the integration tests. However, I would prefer more unit tests and less integration tests - https://martinfowler.com/articles/practical-test-pyramid.html

**Review by easen-amp — CHANGES_REQUESTED** [review]

File: src/commands/content-item/export.ts (hunk 4 above)
Comment (rs-amp) [reply src/commands/content-item/export.ts:200]: ContentItem does not define lastPublishedVersion.

File: src/common/content-item/amplience-schema-validator.ts (hunk 5 above)
Comment (rs-amp) [reply src/common/content-item/amplience-schema-validator.ts:45]: It's used to provide warnings to a user that content being exported will not import correctly with the currently synced content types/schemas. Since you can sync a schema and make older content items invalid, it could be useful for identifying and troubleshooting content items that have become invalid as a result.

It's also used to check if setting missing dependancies as null causes a content item to be invalid. If it does, we ask the user to skip the content entirely instead.

Comment (rs-amp) [reply src/common/content-item/amplience-schema-validator.ts:45]: Draft 7 is supported by default, draft 4 is added manually since I noticed some schemas used it. It seems to be working fine for all content currently in willow and our testing account, but since it doesn't make any _important_ decisions with the validator, I don't think it needs to be lock step with changes to the serverside validation.

File: src/common/content-item/content-dependancy-tree.ts (hunk 6 above)
Comment (rs-amp) [reply src/common/content-item/content-dependancy-tree.ts:43]: That is true, content dependancy logic is tested mostly by the export and import class tests. I can add tests for this class on its own, but it will take some time.

File: src/commands/content-item/export.ts (hunk 4 above)
Comment (easen-amp) [reply src/commands/content-item/export.ts:200]: Ok, in that case my preference would be to add/expose the `lastPublishedVersion` property on the `ContentItem` class in the DC Management SDK

File: src/common/content-item/content-dependancy-tree.ts (hunk 6 above)
Comment (easen-amp) [reply src/common/content-item/content-dependancy-tree.ts:43]: I think that would be useful as you might catch an edge case that your integration tests have missed.

**Comment by rs-amp** [comment]
I've introduced tests for content-dependancy-tree, which includes the following changes:

- Dependancy levels are now strict: before two items would be in the same level if the dependancy was processed before the dependant. This is fine when processing levels synchronously and in order, but that assumption might be broken in future.
- Some renames and cleanup in the content-dependancy-tree class.
- `PublishedContentItem` interface introduced. This includes lastPublishedVersion and lastPublishedDate. In future these should be added to dc-management-sdk.

**Review by easen-amp — APPROVED** [review]
